rts/motoko-rts/src/region.rs

```diff
@@ -0,0 +1,415 @@
+
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub struct BlockId(pub u16);
+
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub struct RegionId(pub u16);
+
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub struct RegionSizeInPages(pub u64);
+
+#[derive(Clone)]
+pub struct AccessVector(pub *mut Blob);
+
+#[derive(Clone)]
+pub struct RegionObject(pub *mut Region);
+
+pub const NIL_REGION_ID : u16 = 0;
+
+// This impl encapsulates encoding of optional region IDs within a u16.
+// Used by block-region table to encode the (optional) region ID of a block.
+impl RegionId {
+    pub fn id_is_nil(id : u16) -> bool {
+	id == NIL_REGION_ID
+    }
+    pub fn from_id(id: u16) -> Self {
+	RegionId(id)
+    }
+    pub fn from_u16(id: u16) -> Option<Self> {
+	if Self::id_is_nil(id) { None } else { Some(RegionId(id - 1)) }
+    }
+    pub fn into_u16(opreg: Option<RegionId>) -> u16 {
+	match opreg {
+	    None => 0,
+	    Some(id) => id.0 + 1
+	}
+    }
+}
+
+// This impl encapsulates encoding of optional region (sizes) within a u64.
+// Used by region table to encode sizes of allocated regions (the size of which are Some(s)),
+// and which regions are available to allocate (the size of which are None).
+impl RegionSizeInPages {
+    pub fn from_size_in_pages(s: u64) -> Self {
+	RegionSizeInPages(s)
+    }
+    pub fn u64_is_nil(u : u64) -> bool {
+	u == 0
+    }
+    pub fn from_u64(u: u64) -> Option<Self> {
+	if Self::u64_is_nil(u) { None } else { Some(RegionSizeInPages(u - 1)) }
+    }
+    pub fn into_u64(opreg: Option<RegionSizeInPages>) -> u64 {
+	match opreg {
+	    None => 0,
+	    Some(s) => s.0 + 1
+	}
+    }
+}
+
+impl AccessVector {
+
+    pub fn from_value(v: &Value) -> Self {
+	AccessVector(v.get_ptr() as *mut Blob)
+    }
+
+    pub unsafe fn set_ith_block_id(&self, i: u32, block_id: &BlockId) {
+	let block_id_upper : u8 = ((block_id.0 & 0xff00) >> 8) as u8;
+	let block_id_lower : u8 = ((block_id.0 & 0x00ff) >> 0) as u8;
+
+	// Update heap memory with new association.
+	// (write big-endian u16 to slot i in new_pages.)
+	self.0.set(i * 2 + 0, block_id_upper);
+	self.0.set(i * 2 + 1, block_id_lower);
+    }
+
+    pub unsafe fn get_ith_block_id(&self, i: u32) -> BlockId {
+	let upper : u16 = self.0.get(i * 2 + 0) as u16;
+	let lower : u16 = self.0.get(i * 2 + 1) as u16;
+	BlockId(upper << 8 | lower)
+    }
+}
+
+impl RegionObject {
+
+    pub unsafe fn from_value(v: &Value) -> Self {
+	RegionObject(v.as_region())
+    }
+
+    pub unsafe fn local_offset_into_global_offset(&self, offset: u64) -> u64 {
+	let av = AccessVector::from_value(&(*self.0).vec_pages);
+
+	// assert that offset is currently allocated
+	// (to do -- this should be a certain kind of trap when it fails.)
+	assert!(offset / meta_data::size::PAGE_IN_BYTES < (*self.0).page_count.into());
+
+	// Which block (rank relative to this region)?
+	let block_rank  = offset / meta_data::size::BLOCK_IN_BYTES;
+
+	// Where in that block?
+	let intra_block_index = offset % meta_data::size::BLOCK_IN_BYTES;
```

Comment:
I wonder if you can do this in one operation (div_rem?) eventually. No need now...

Reply:
Actually, looks like Rust doesn't support div_rem and you can rely on the compiler optimizing the two computations (also, not necessary now anyway)